Apply RBAC with `casbin` package for access control

peterxcli opened the pull request:
use casbin to add authorize feature

The system is a library with 3 roles: Director, Librarian, and Visitor.
Everyone can find all the books in the library.
registered visitors can borrow books, return books, and view their own borrowing records.
librarians can update the information of books and have all the privileges of registered visitors.
The director has full access to the librarians and can delete books. Deleting a book is a soft delete, so it does not affect the borrowing record.

peterxcli committed:
- Implement role-based access control (RBAC) with Casbin
- Add Casbin middleware for RBAC to book and record routes
- Add role field to Identity and User domain models
- Update book and record route registration to accept an Application parameter
- Initialize Enforcer in the application bootstrap process
- Update dependencies to include Casbin

peterxcli committed:
- Add new dependencies to go.mod
- Modify enforcer.go to use project root directory for file paths
- Create enforcer_test.go to test authorization rules

peterxcli committed:
add test target for makefile

peterxcli committed:
- Remove commented out code for getting project directory and safely joining paths in `rbac/enforcer.go` and `rbac/enforcer_test.go`

alan890104 commented:
Amazing bro. Please give me several days to review these changes.

alan890104 commented:
@peterxcli BTW, could you please add these scenarios into `README.md`?

> The system is a library with 3 roles: Director, Librarian, and Visitor.
Everyone can find all the books in the library.
registered visitors can borrow books, return books, and view their own borrowing records.
librarians can update the information of books and have all the privileges of registered visitors.
The director has full access to the librarians and can delete books. Deleting a book is a soft delete, so it does not affect the borrowing record.

peterxcli committed:
add Scenarios into README

peterxcli commented:
> @peterxcli BTW, could you please add these scenarios into `README.md`?
> 
> > The system is a library with 3 roles: Director, Librarian, and Visitor.
> > Everyone can find all the books in the library.
> > registered visitors can borrow books, return books, and view their own borrowing records.
> > librarians can update the information of books and have all the privileges of registered visitors.
> > The director has full access to the librarians and can delete books. Deleting a book is a soft delete, so it does not affect the borrowing record.

done

alan890104 committed:
chore: update `.env.example` add missing `LIBRARY_DB_KIND`

alan890104 committed:
refactor: unify the format of route parameters

alan890104 committed:
refactor: remove unused import

alan890104 committed:
chore: readonly for caddyfile

alan890104 committed:
chore: role definition should be const

alan890104 committed:
fix: named convension

alan890104 commented on rbac/enforcer.go:
```diff
@@ -0,0 +1,26 @@
+
+func NewEnforcer() *casbin.Enforcer {
+	// Get the absolute path to the project directory
+	projectDir, err := projectRootDir.GetRootDirectory()
```
Why is it necessary to introduce this third-party package (which is not widely used) to address the path issue? Could you describe the reasons for its usage?

peterxcli replied:
This is for me to test why I can't read the casbin model and policy file initially when I ran `go test ./...`, I will removed it.
Thanks for the reminder!

alan890104 reviewed (changes requested):
I have refactored a little bit part in your code, especially routes. Additionally, could you please test these APIs in `stage` mode? I believe there may be some path issues that need to be addressed. (e.g. mount rbac policy to volume)

peterxcli committed:
chore: Simplify file path construction for casbin file

peterxcli committed:
chore: go mod tidy

peterxcli committed:
chore: copy casbin file from build stage to runtime stage in Dockerfile

peterxcli commented:
> I have refactored a little bit part in your code, especially routes. Additionally, could you please test these APIs in `stage` mode? I believe there may be some path issues that need to be addressed. (e.g. mount rbac policy to volume)

resolved the path issue by copying the casbin config into runtime stage, the api access for `stage` built container recover

<img width="509" alt="Screenshot 2023-10-25 at 2 39 54 AM" src="https://github.com/alan890104/go-clean-architecture/assets/54034701/ec9c2850-0c20-4a9a-b633-8b74db15b0e0">

alan890104 commented on docker/Dockerfile:
```diff
@@ -25,5 +25,9 @@ FROM busybox AS runtime
 # Copy the binary from the build stage to the runtime stage
 COPY --from=build /app/app /app
 
+# Copy the configuration files from the build stage to the runtime stage
```
This method works as well, but I prefer to follow another design principle: `app = image + config`.

We can mount `model.conf` and `policy.csv` using Docker Compose volumes, so that whenever we change the configuration files, we don't have to rebuild the image again.

peterxcli replied:
I get it

alan890104 commented on rbac/enforcer.go:
```diff
@@ -5,18 +5,11 @@ import (
-
-	modelPath := filepath.Join(projectDir, "rbac", "model.conf")
-	policyPath := filepath.Join(projectDir, "rbac", "policy.csv")
+	modelPath := filepath.Join("rbac", "model.conf")
```
This is now more elegant than it was previously :)

peterxcli committed:
- Remove copying of RBAC configuration files in Dockerfile
- Add RBAC configuration files as read-only volumes in docker-compose.stage.yaml

alan890104 commented on docker/docker-compose.stage.yaml:
```diff
@@ -34,6 +34,8 @@ services:
     env_file:
       - $PWD/.env.stage
+    volumes:
+      - $PWD/rbac:/rbac:ro
```
maybe we can mount those configuration files instead of mounting whole folders?

peterxcli replied:
ur right. thx

peterxcli committed:
chore: Mount RBAC configuration files separately instead of the whole folder

alan890104 reviewed (approved):
LGTM
